### DISCUSSION TITLE
logging in Operator not work

### DISCUSSION THREAD

**Discussion by SamJia** [apache/airflow#21021] [General]
### Apache Airflow version

2.2.3 (latest released)

### What happened

Operator logging not work. 
For example, for a task with logging:
![image](https://user-images.githubusercontent.com/12780802/150499992-d939cdfc-0e3f-471e-b3ad-7232b8f313d9.png)
the log in webserver is:
![image](https://user-images.githubusercontent.com/12780802/150500091-e357c1b8-8cab-424f-aed1-21c957c6d6ed.png)
no any log information are recorded.

In my view, the reason is that when a TaskInstance is loaded through session with DagRun.run_id in function `_get_ti` in airflow/cli/commands/task_command.py, ti._log object is not properly initialzed.
![image](https://user-images.githubusercontent.com/12780802/150521672-0a9fd2d6-c231-4b9a-812e-78910c067bd4.png)

Then, when call 'ti.init_run_context', ti._log will be set as logging.getLogger(self.__class__.__module__ + '.' + self.__class__.__name__) according to function log of class LoggingMixin in airflow/utils/log/logging_mixin.py. And the context of logging.getLogger('airflow.task') is not properly set.
![image](https://user-images.githubusercontent.com/12780802/150521756-1816cde0-9616-48d6-aca9-e18203e21f1b.png)

Following, in function '_capture_task_logs' in airflow/cli/commands/task_command.py, the root logger configuration is replaced with the airflow.task configuration, which context is not set.
![image](https://user-images.githubusercontent.com/12780802/150521822-4c6a8840-2a60-459b-a063-2b22a0706295.png)

It means all the loging happen after the replacement cannot be recorded.

Possible way to fix:

In function `refresh_from_task`, initialize self._log like follows:
![image](https://user-images.githubusercontent.com/12780802/150521531-3b53c984-1a81-40da-8b04-d99babe1de54.png)

Moreover, it may be also addeed in function `refresh_from_db`

### What you expected to happen

I wish the log looks like this:
![image](https://user-images.githubusercontent.com/12780802/150502087-9b8dd557-f60a-45fe-a179-9223b32dfbba.png)

### How to reproduce

My config related to logging in airlfow.cfg:
```
[logging]
base_log_folder = /home/airflow/airflow/logs
remote_logging = False
remote_log_conn_id =
google_key_path =
remote_base_log_folder =
encrypt_s3_logs = False
logging_level = INFO
fab_logging_level = INFO
logging_config_class =
colored_console_log = True
colored_log_format = [%%(blue)s%%(asctime)s%%(reset)s] {%%(blue)s%%(filename)s:%%(reset)s%%(lineno)d} %%(log_color)s%%(levelname)s%%(reset)s - %%(log_color)s%%(message)s%%(reset)s
colored_formatter_class = airflow.utils.log.colored_log.CustomTTYColoredFormatter
log_format = [%%(asctime)s] {%%(filename)s:%%(lineno)d} %%(levelname)s - %%(message)s
simple_log_format = %%(asctime)s %%(levelname)s - %%(message)s
task_log_prefix_template =
log_filename_template = {{ ti.dag_id }}/{{ ti.task_id }}/{{ ts }}/{{ try_number }}.log
log_processor_filename_template = {{ filename }}.log
dag_processor_manager_log_location = /home/airflow/airflow/logs/dag_processor_manager/dag_processor_manager.log
task_log_reader = task
extra_logger_names =
worker_log_server_port = 8793
```

My DAG and task:

```
import datetime

from airflow import DAG
from airflow.decorators import task
from airflow.operators.python import PythonOperator


def test_log():
    import logging

    logging.info("airflow.task >>> INFO logger test")
    logging.warn("airflow.task >>> WARN logger test")
    logging.error("airflow.task >>> ERROR logger test")
    logging.critical("airflow.task >>> CRITICAL logger test")

with DAG(
    'test_log',
    description='test log',
    schedule_interval=None,
    start_date=datetime.datetime(2022, 1, 19),
    catchup=False,
    tags=['log'],
) as dag:
    task = PythonOperator(
        task_id='test_log_task',
        python_callable=test_log,
        dag=dag
    )
```

### Operating System

NAME="Ubuntu" VERSION="18.04.4 LTS (Bionic Beaver)" ID=ubuntu ID_LIKE=debian PRETTY_NAME="Ubuntu 18.04.4 LTS" VERSION_ID="18.04" HOME_URL="https://www.ubuntu.com/" SUPPORT_URL="https://help.ubuntu.com/" BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/" PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy" VERSION_CODENAME=bionic UBUNTU_CODENAME=bionic

### Versions of Apache Airflow Providers

None

### Deployment

Virtualenv installation

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template!

**Comment by potiuk** [comment]
You are using logging module improperly. By `logging,info` (and similar) you are logging stuff to "root" logger:

https://docs.python.org/3/library/logging.html#logging.info

> logging.info(msg, *args, **kwargs)¶
Logs a message with level INFO on the root logger. The arguments are interpreted as for debug().

Instead you should log message to "airflow.task" logger if you want messages to show up in task log:

```
logger = logging.getLogger("airflow.task")
logger.info(...)
``

**Comment by SamJia** [comment]
> You are using logging module improperly. By `logging,info` (and similar) you are logging stuff to "root" logger:
> 
> https://docs.python.org/3/library/logging.html#logging.info
> 
> > logging.info(msg, *args, **kwargs)¶
> > Logs a message with level INFO on the root logger. The arguments are interpreted as for debug().
> 
> Instead you should log message to "airflow.task" logger if you want messages to show up in task log:
> 
> ```
> logger = logging.getLogger("airflow.task")
> logger.info(...)
> ``
> ```


Actually I have tried to use logger `airflow.task`, but also failed

![image](https://user-images.githubusercontent.com/12780802/150556056-90dfae46-1352-434a-a455-8668289d58c9.png)
![image](https://user-images.githubusercontent.com/12780802/150556096-640b9487-50c4-4691-82fd-0d18606d9ca3.png)

Moreover, considering the level and handler of `root` logger are replaced with the level and handler of `airflow.task` logger, I think they should work same in task operators.

**Comment by potiuk** [comment]
Cannot reproduce it with "standard" airflow configuration. You must have some custom logging configuration in local settings that causes it:


```python
import datetime

from airflow import DAG
from airflow.decorators import task
from airflow.operators.python import PythonOperator


def test_log():
    import logging

    logger = logging.getLogger("airflow.task")
    logger.info("airflow.task >>> INFO logger test")
    logger.warn("airflow.task >>> WARN logger test")
    logger.error("airflow.task >>> ERROR logger test")
    logger.critical("airflow.task >>> CRITICAL logger test")

with DAG(
    'test_log',
    description='test log',
    schedule_interval=None,
    start_date=datetime.datetime(2022, 1, 19),
    catchup=False,
    tags=['log'],
) as dag:
    task = PythonOperator(
        task_id='test_log_task',
        python_callable=test_log,
        dag=dag
    )

```

Resulting log:
```

*** Reading local file: /root/airflow/logs/test_log/test_log_task/2022-01-21T16:17:12.235643+00:00/1.log
[2022-01-21, 16:17:12 UTC] {taskinstance.py:1032} INFO - Dependencies all met for <TaskInstance: test_log.test_log_task manual__2022-01-21T16:17:12.235643+00:00 [queued]>
[2022-01-21, 16:17:12 UTC] {taskinstance.py:1032} INFO - Dependencies all met for <TaskInstance: test_log.test_log_task manual__2022-01-21T16:17:12.235643+00:00 [queued]>
[2022-01-21, 16:17:12 UTC] {taskinstance.py:1238} INFO - 
--------------------------------------------------------------------------------
[2022-01-21, 16:17:12 UTC] {taskinstance.py:1239} INFO - Starting attempt 1 of 1
[2022-01-21, 16:17:12 UTC] {taskinstance.py:1240} INFO - 
--------------------------------------------------------------------------------
[2022-01-21, 16:17:12 UTC] {taskinstance.py:1259} INFO - Executing <Task(PythonOperator): test_log_task> on 2022-01-21 16:17:12.235643+00:00
[2022-01-21, 16:17:12 UTC] {standard_task_runner.py:52} INFO - Started process 465 to run task
[2022-01-21, 16:17:12 UTC] {standard_task_runner.py:76} INFO - Running: ['***', 'tasks', 'run', 'test_log', 'test_log_task', 'manual__2022-01-21T16:17:12.235643+00:00', '--job-id', '3', '--raw', '--subdir', 'DAGS_FOLDER/test.py', '--cfg-path', '/tmp/tmpjrm_lflu', '--error-file', '/tmp/tmpowwm0hd3']
[2022-01-21, 16:17:12 UTC] {standard_task_runner.py:77} INFO - Job 3: Subtask test_log_task
[2022-01-21, 16:17:12 UTC] {logging_mixin.py:109} INFO - Running <TaskInstance: test_log.test_log_task manual__2022-01-21T16:17:12.235643+00:00 [running]> on host afc5ced1b873
[2022-01-21, 16:17:12 UTC] {taskinstance.py:1426} INFO - Exporting the following env vars:
AIRFLOW_CTX_DAG_OWNER=***
AIRFLOW_CTX_DAG_ID=test_log
AIRFLOW_CTX_TASK_ID=test_log_task
AIRFLOW_CTX_EXECUTION_DATE=2022-01-21T16:17:12.235643+00:00
AIRFLOW_CTX_DAG_RUN_ID=manual__2022-01-21T16:17:12.235643+00:00
[2022-01-21, 16:17:12 UTC] {test.py:12} INFO - ***.task >>> INFO logger test
[2022-01-21, 16:17:12 UTC] {test.py:13} WARNING - ***.task >>> WARN logger test
[2022-01-21, 16:17:12 UTC] {test.py:14} ERROR - ***.task >>> ERROR logger test
[2022-01-21, 16:17:12 UTC] {test.py:15} CRITICAL - ***.task >>> CRITICAL logger test
[2022-01-21, 16:17:12 UTC] {python.py:175} INFO - Done. Returned value was: None
[2022-01-21, 16:17:12 UTC] {taskinstance.py:1277} INFO - Marking task as SUCCESS. dag_id=test_log, task_id=test_log_task, execution_date=20220121T161712, start_date=20220121T161712, end_date=20220121T161712
[2022-01-21, 16:17:12 UTC] {local_task_job.py:154} INFO - Task exited with return code 0
[2022-01-21, 16:17:12 UTC] {local_task_job.py:264} INFO - 0 downstream tasks scheduled from follow-on schedule check

```

**Reply by potiuk** [reply]
Converted it into discussion. This ^^ is plain Airflow 2.2.3 with default configuration. Please take a look at your logging configuration and see what changes you added (ideally difff) of the standard configuration.

**Reply by potiuk** [reply]
And I confirm that plain `logging.info` works as well. TIL.

**Reply by potiuk** [reply]
BTW. Are you using default impersonation maybe @SamJia?  https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html#default-impersonation - that might be another reason for this - see #19376

**Reply by ArgentFalcon** [reply]
So we're running into this issue as well, but not using impersonation. We are running the celery workers on top of kubernetes, but that shouldn't be changing the execution pattern. I'm digging deeper into this, but I have tested this with default logging configuration on our cluster, and confirmed it doesn't work, so I think it must be more intrinsic to how we're running Airflow, rather than just a configuration change. I don't know if it affects Airflow 2.3 though yet (we're not there), but when we upgrade, I'll do further testing.

**Comment by WALEX2000** [comment]
Hey!  Have you found a solution?
I have the exact same problem and nothing seems to work

**Reply by SamJia** [reply]
I fix it by adding some codes in the installed package coding files.

In function `refresh_from_task` of file `airflow/models/taskinstance.py`, add one line to initialize self._log like follows:
![150521531-3b53c984-1a81-40da-8b04-d99babe1de54](https://user-images.githubusercontent.com/12780802/158189944-c1043ecc-1853-4879-9814-a1a685924691.png)

Moreover, it may be also addeed in function `refresh_from_db`

**Reply by potiuk** [reply]
Question - that has not been responded so far - are all the people who experience it using impersonation? This issue is not reproducible in "normal" circumstances so it would be good to get to the bottom of it.

**Reply by WALEX2000** [reply]
I'm not using impersonation. This was also only happening with the BashOperator task. But I then realised I can just use normal prints to display the information I want on. the BashOperator, so it's not an issue for me anymore.

**Comment by PowerToThePeople111** [comment]
We are experiencing the same problems. I tried several ways to log information in a virtualenv-operator: by using print-statements, logging.info, and the special logger ( logging.getLogger("airflow.task") ). None of them worked for us. 

We are using airflow in a kubernetes setup. ( very similar to this issue https://github.com/apache/airflow/issues/34865 )

Is the fix provided by SamJia (adding the red-marked lined to the `refresh_from_task` function ) expected to work with the latest version? I had a look into the code and it seemed like the function has been changed.

Given what I read here and in the other issue, i got the feeling that this might be related to the config somehow: in your "standart config" this seems to work for you but other people had problems. Do you have any idea which config properties might be the culprits here?

**Reply by PowerToThePeople111** [reply]
Ok, we got it to work by changing the info logs to warn logs:

So every time, we want to log something, we call:
```python
def log_info(message):
    logging.getLogger("airflow.task").warning(message)
```

From our guy in the infra team i know that we do remote logging with log level set to info.


The logs which are being produced show those warnings as infos:
[2024-11-07, 12:59:20 UTC] {process_utils.py:191} INFO - XYZ
[2024-11-07, 12:59:26 UTC] {process_utils.py:191} INFO - XYZ
[2024-11-07, 12:59:33 UTC] {process_utils.py:191} INFO - XYZ
[2024-11-07, 12:59:40 UTC] {process_utils.py:191} INFO - XYZ

So this workaround is doing it for us.
